pkg/sql/logictest/tests/cockroach-go-testserver-23.1/cockroach-go-testserver-23_1_test: TestLogic_upgrade_skip_version failed

pkg/sql/logictest/tests/cockroach-go-testserver-23.1/cockroach-go-testserver-23_1_test.TestLogic_upgrade_skip_version [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13652588?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13652588?buildTab=artifacts#/) on master @ [326f0dfc8f9ed60f083720eb021267669df5b630](https://github.com/cockroachdb/cockroach/commits/326f0dfc8f9ed60f083720eb021267669df5b630):


```
=== RUN   TestLogic_upgrade_skip_version
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/c7caea9876d3f2febbad5ba87e564e67/logTestLogic_upgrade_skip_version2030284696
    test_log_scope.go:81: use -show-logs to present logs inline
    logic.go:2910: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11346/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-23.1/cockroach-go-testserver-23_1_test_/cockroach-go-testserver-23_1_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/upgrade_skip_version:15: SELECT crdb_internal.release_series(version) FROM [SHOW CLUSTER SETTING version]
        expected:
            24.1
        but found (query options: "retry") :
            23.2
            
[21:54:15] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11346/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-23.1/cockroach-go-testserver-23_1_test_/cockroach-go-testserver-23_1_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/upgrade_skip_version: 2 statements
[21:54:15] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11346/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/cockroach-go-testserver-23.1/cockroach-go-testserver-23_1_test_/cockroach-go-testserver-23_1_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/upgrade_skip_version with config cockroach-go-testserver-23.1: 2 tests, 1 failures
cockroach logs captured in: /artifacts/tmp/_tmp/c7caea9876d3f2febbad5ba87e564e67/cockroach-logs1183877457
[21:54:17] --- total progress: 2 statements
--- total: 2 tests, 1 failures
    logic.go:4186: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c7caea9876d3f2febbad5ba87e564e67/logTestLogic_upgrade_skip_version2030284696
--- FAIL: TestLogic_upgrade_skip_version (227.82s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_upgrade_skip_version.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35572

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -306,11 +306,12 @@ import (
 //      - colnames: column names are verified (the expected column names
 //            are the first line in the expected results).
 //      - retry: if the expected results do not match the actual results, the
-//            test will be retried with exponential backoff up to some maximum
-//            duration. If the test succeeds at any time during that period, it
-//            is considered successful. Otherwise, it is a failure. See
-//            testutils.SucceedsSoon for more information. If run with the
-//            -rewrite flag, the query will be run only once after a 2s sleep.
+//            test will be retried with exponential backoff up to the duration
+//            given by retry_duration. If the test succeeds at any time during
+//            that period, it is considered successful. Otherwise, it is a
+//            failure. See testutils.SucceedsSoon for more information. If run
+//            with the -rewrite flag, the query will be run only once after a
+//            2s sleep.
 //      - async: runs the query asynchronously, marking it as a pending
 //            query using the label parameter as a unique name, to be completed
 //            and validated later with "awaitquery". This is intended for use
@@ -417,6 +418,10 @@ import (
 //    (including those which expect errors) will be retried for a fixed
 //    duration until the test passes, or the alloted time has elapsed.
 //    This is similar to the retry option of the query directive.
+//
+//  - retry_duration <duration>
+//    Specifies the amount of time to retry when using the retry directive.
+//    Defaults to testutils.DefaultSucceedsSoonDuration (45 seconds).
 //
 // The overall architecture of TestLogic is as follows:
 //
@@ -1080,6 +1085,10 @@ type logicTest struct {
 	// to false after every successful statement or query test point, including
 	// those which are supposed to error out.
 	retry bool
+
+	// retryDuration is the maximum duration to retry a statement when using
+	// the retry directive.
+	retryDuration time.Duration
 }
 
 func (t *logicTest) t() *testing.T {
@@ -1876,6 +1885,7 @@ func (t *logicTest) setup(
 	tempExternalIODir, tempExternalIODirCleanup := testutils.TempDir(t.rootT)
 	t.sharedIODir = tempExternalIODir
 	t.testCleanupFuncs = append(t.testCleanupFuncs, tempExternalIODirCleanup)
+	t.retryDuration = testutils.DefaultSucceedsSoonDuration
 
 	if cfg.UseCockroachGoTestserver {
 		skip.UnderRace(t.t(), "test uses a different binary, so the race detector doesn't work")
@@ -2483,6 +2493,21 @@ func (t *logicTest) processSubtest(
 				)
 			}
 			repeat = count
+
+		case "retry_duration":
+			var duration time.Duration
+			var err error
+			if len(fields) != 2 {
+				err = errors.New("invalid line format")
+			} else {
+				duration, err = time.ParseDuration(fields[1])
+			}
+			if err != nil {
+				return errors.Wrapf(err, "%s:%d invalid retry_duration line",
+					path, s.Line+subtest.lineLineIndexIntoFile)
+			}
+			t.retryDuration = duration
+
 		case "skip_on_retry":
 			t.skipOnRetry = true
 
@@ -2567,12 +2592,12 @@ func (t *logicTest) processSubtest(
 					var cont bool
 					var err error
 					if t.retry {
-						err = testutils.SucceedsSoonError(func() error {
+						err = testutils.SucceedsWithinError(func() error {
 							t.purgeZoneConfig()
 							var tempErr error
 							cont, tempErr = t.execStatement(stmt)
 							return tempErr
-						})
+						}, t.retryDuration)
 					} else {
 						cont, err = t.execStatement(stmt)
 					}
@@ -2903,10 +2928,10 @@ func (t *logicTest) processSubtest(
 
 				for i := 0; i < repeat; i++ {
 					if t.retry && !*rewriteResultsInTestfiles {
-						if err := testutils.SucceedsSoonError(func() error {
+						if err := testutils.SucceedsWithinError(func() error {
 							t.purgeZoneConfig()
 							return t.execQuery(query)
-						}); err != nil {
+						}, t.retryDuration); err != nil {
 							t.Error(err)
 						}
 					} else {

--- a/pkg/sql/logictest/testdata/logic_test/upgrade_skip_version
+++ b/pkg/sql/logictest/testdata/logic_test/upgrade_skip_version
@@ -9,7 +9,7 @@ upgrade all
 
 # We have seen that upgrades can take a long time in CI. Give some extra time
 # for the upgrade to complete.
-sleep 30s
+retry_duration 10m
 
 # Verify that the cluster is upgrading to 24.1.
 query T retry

